components/html-block/html-block.js

```diff
@@ -155,15 +171,51 @@ class HtmlBlock extends LitElement {
 
 		this.shadowRoot.innerHTML += '<div class="d2l-html-block-rendered"></div><slot></slot>';
 
+		this.shadowRoot.querySelector('slot').addEventListener('slotchange', async e => {
+
+			const template = e.target.assignedNodes({ flatten: true })
+				.find(node => (node.nodeType === Node.ELEMENT_NODE && node.tagName === 'TEMPLATE'));
+
+			this._stamp(template);
+
+		});
+		this._renderContainer = this.shadowRoot.querySelector('.d2l-html-block-rendered');
+		this._context = { ...this._contextObserverController.values };
+	}
+
+	updated() {
+		super.updated();
+		if (this._contextObjectHasChanged()) {
+			const template = this.querySelector('template');
+			this._stamp(template);
+		}
+	}
+
+	_contextObjectHasChanged() {
+		if (this._context.size !== this._contextObserverController.values.size) return true;
+		for (const [attr, val] of this._context) {
+			if (!this._contextObserverController.values.has(attr)) return true;
+			if (this._contextObserverController.values.get(attr) !== val) return true;
+		}
+		return false;
+	}
+
+	_stamp(template) {
 		const stampHTML = async template => {
 			const fragment = template ? document.importNode(template.content, true) : null;
 			if (fragment) {
 
 				let temp = document.createElement('div');
 				temp.appendChild(fragment);
 
-				for (const render of HtmlBlock._renderers) {
-					temp = await render(temp);
+				for (const renderer of getRenderers()) {
+					if (renderer.contextAttributes) {
+						const contextValues = new Map();
+						renderer.contextAttributes.forEach(attr => contextValues.set(attr, this._contextObserverController.values.get(attr)));
```

Comment:
Maybe there is a way we can avoid having to do this for every instance of a `d2l-html-block`...?  I hesitate to provide the entire context though.

Reply:
So the mutation observer underlying the controller is shared, which is good, but if we did pass the entire context to each renderer we would avoid needing to do this mapping. I think filtering the map needs to be done either here or in the renderer, but I don't have a strong feeling of which is better.

Reply:
There is a single set of renderers in this module outside of the class.  We could keep the values for the attributes per renderer, either in this module or on the renderer which is shared across instances of html-block.  The difference is probably negligible even when there are many html-blocks though, since the collection of attributes should be so small.  If it weren't, I think avoiding the nested looping would be desirable.

Reply:
I'm not sure I understand what you mean. If we pass the full context object to the renderer, it can be responsible for pulling out its attributes. But ultimately the controller itself must be specific to each html-block.

Reply:
Yep, we could pass them all, even though it didn't ask us to.  I am just saying that we could avoid this looping, either by passing them all, or caching the set of values per renderer outside the component so we only need to do it once. It would be a lookup in a Map vs looping over the entire collection for each html-block. But as mentioned, as long as the number of attributes is small, the work should be negligible.

Reply:
Ah okay, caching is an interesting option. I think that's probably better than passing the whole thing. I think you're right that this should be negligible because it's always possible to condense attributes if things start to get unreasonable. That said, I do like the idea of maintaining a map outside of the component that we can reference and update this way.

I think we can also get some minor efficiencies this way because the mutation observer will fire if I do something like this: `elem.setAttribute('attr', elem.getAttribute('attr'))` even though it's clearly not going to change anything. So there's probably some efficiency to be had in not triggering a re-render unless that mapped value has actually changed.